src/main/scala/decider/Z3ProverAPI.scala

```diff
@@ -0,0 +1,439 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2021 ETH Zurich.
+
+package viper.silicon.decider
+
+import com.typesafe.scalalogging.LazyLogging
+import viper.silicon.common.config.Version
+import viper.silicon.interfaces.decider.{Prover, Result, Sat, Unknown, Unsat}
+import viper.silicon.state.IdentifierFactory
+import viper.silicon.state.terms.{App, Decl, Fun, FunctionDecl, Implies, MacroDecl, Sort, SortDecl, SortWrapperDecl, Term, sorts}
+import viper.silicon.{Config, Map}
+import viper.silicon.verifier.Verifier
+import viper.silver.reporter.Reporter
+import viper.silver.verifier.{ApplicationEntry, ConstantEntry, MapEntry, ModelEntry, ModelParser, ValueEntry, DefaultDependency => SilDefaultDependency, Model => ViperModel}
+import java.io.PrintWriter
+import java.nio.file.Path
+
+import scala.collection.mutable
+import com.microsoft.z3._
+import viper.silicon.reporting.ExternalToolError
+
+import scala.collection.immutable.ListMap
+import scala.jdk.CollectionConverters.MapHasAsJava
+import scala.util.Random
+
+
+object Z3ProverAPI {
+  val name = "Z3-API"
+  val minVersion = Version("4.8.6.0")
+  val maxVersion = Some(Version("4.8.6.0")) /* X.Y.Z if that is the *last supported* version */
+  val exeEnvironmentalVariable = "Z3_EXE"
+  val dependencies = Seq(SilDefaultDependency("Z3", minVersion.version, "https://github.com/Z3Prover/z3"))
+  val staticPreamble = "/z3config.smt2"
+  val startUpArgs = Seq("-smt2", "-in")
+  val randomizeSeedsOptions = Seq("random_seed")
+  val randomizeSeedsSetting = "RANDOMIZE_SEEDS"
+  val initialOptions = Map("auto_config" -> "false", "type_check" -> "true")
```

Comment:
Would be good to comment what the various flags do for future reference? Do we know, or are they copied over from an older codebase?

Reply:
They are copied from the Z3 options in z3config.smt2, and I removed the options that no longer work (since they result in ugly error messages). I'll add a comment.